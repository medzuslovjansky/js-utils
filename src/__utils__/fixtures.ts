type FixtureGetter = () => [string, string, string, string][];
type StringsGetter = () => string[];

export const adjectives: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/adjectives.json');
export const nouns_feminine: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-feminine.json');
export const nouns_masculine_animate: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-masculine-animate.json');
export const nouns_masculine: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-masculine.json');
export const nouns_misc: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-misc.json');
export const nouns_neuter: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-neuter.json');
export const verbs_imperfect: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/verbs-imperfect.json');
export const verbs_perfect: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/verbs-perfect.json');
export const verbs_misc: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/verbs-misc.json');
export const pronouns_demonstrative: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-demonstrative.json');
export const pronouns_indefinite: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-indefinite.json');
export const pronouns_interrogative: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-interrogative.json');
export const pronouns_personal: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-personal.json');
export const pronouns_possessive: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-possessive.json');
export const pronouns_reciprocal: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-reciprocal.json');
export const pronouns_reflexive: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-reflexive.json');
export const pronouns_relative: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-relative.json');
export const other: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/other.json');
export const word_forms: StringsGetter = () =>
  jest.requireActual('../__fixtures__/word-forms.json');
