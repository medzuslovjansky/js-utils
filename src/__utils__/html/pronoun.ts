/* eslint-disable prettier/prettier */
import type { SteenPronounParadigm } from '../../pronoun';
import * as common from './common';

export function render(value: SteenPronounParadigm | null) {
  if (!value) {
    return '';
  }

  const template = value.type === 'noun' ? noun : adjective;
  return template(common.get(value));
}

export function renderDiff(a: SteenPronounParadigm | null, b: SteenPronounParadigm | null) {
  if (a && b && a.type !== b.type) {
    throw new Error(`Pronoun types do not match (${a.type} !== ${b.type})`);
  }

  const type = a?.type ?? b?.type;
  if (!type) {
    return '';
  }

  const template = type === 'noun' ? noun : adjective;
  return template(common.diff(a, b));
}

function noun(r: (key: keyof SteenPronounParadigm, subkey: string, index?: number) => string | null): string {
  const figure = `\
<figure>
<figcaption>Pronoun declension</figcaption>
<table>
  <thead>
    <tr>
      <th>case</th>
      <th>Word Form</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Nom.</th>
      <td>${r('cases', 'nom', 0)}</td>
    </tr>
    <tr>
      <th>Acc.</th>
      <td>${r('cases', 'acc', 0)}</td>
    </tr>
    <tr>
      <th>Gen.</th>
      <td>${r('cases', 'gen', 0)}</td>
    </tr>
    <tr>
      <th>Loc.</th>
      <td>${r('cases', 'loc', 0)}</td>
    </tr>
    <tr>
      <th>Dat.</th>
      <td>${r('cases', 'dat', 0)}</td>
    </tr>
    <tr>
      <th>Ins.</th>
      <td>${r('cases', 'ins', 0)}</td>
    </tr>
  </tbody>
</table>
</figure>
`;

  return figure;
}

function adjective(r: (key: keyof SteenPronounParadigm, subkey: string, index?: number) => string | null): string {
  const figure = `\
<figure>
  <figcaption>Pronoun declension</figcaption>
  <table>
    <caption>singular</caption>
    <thead>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>neuter</th>
        <th>feminine</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>Nom.</th>
        <td>${r('casesSingular', 'nom', 0)}</td>
        <td>${r('casesSingular', 'nom', 1)}</td>
        <td>${r('casesSingular', 'nom', 2)}</td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td>${r('casesSingular', 'acc', 0)}</td>
        <td>${r('casesSingular', 'acc', 1)}</td>
        <td>${r('casesSingular', 'acc', 2)}</td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2">${r('casesSingular', 'gen', 0)}</td>
        <td>${r('casesSingular', 'gen', 1)}</td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2">${r('casesSingular', 'loc', 0)}</td>
        <td>${r('casesSingular', 'loc', 1)}</td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2">${r('casesSingular', 'dat', 0)}</td>
        <td>${r('casesSingular', 'dat', 1)}</td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2">${r('casesSingular', 'ins', 0)}</td>
        <td>${r('casesSingular', 'ins', 1)}</td>
      </tr>
    </tbody>
  </table>

  <table>
    <caption>plural</caption>
    <tbody>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>feminine/neuter</th>
      </tr>
      <tr>
        <th>Nom.</th>
        <td>${r('casesPlural', 'nom', 0)}</td>
        <td>${r('casesPlural', 'nom', 1)}</td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td>${r('casesPlural', 'acc', 0)}</td>
        <td>${r('casesPlural', 'acc', 1)}</td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2">${r('casesPlural', 'gen', 0)}</td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2">${r('casesPlural', 'loc', 0)}</td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2">${r('casesPlural', 'dat', 0)}</td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2">${r('casesPlural', 'ins', 0)}</td>
      </tr>
    </tbody>
  </table>
</figure>
`;

  return figure;
}
