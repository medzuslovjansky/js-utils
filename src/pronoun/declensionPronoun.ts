import { declensionAdjective } from '../adjective';
import { stripDiacritics } from '../common';
import transliterate from '../transliterate';
import {
  ALONE,
  FIRST_PERSON,
  POSSESSIVE,
  REFLEXIVE,
  SECOND_PERSON,
  THIRD_PERSON,
} from './lookups';

/** @deprecated */
export interface SteenPronounParadigm {
  type: string;
  columns?: string[];
  cases?: Record<string, any>;
  casesSingular?: Record<string, any>;
  casesPlural?: Record<string, any>;
}

export function declensionPronounFlat(rawWord: string, pronounType: string) {
  const paradigm = declensionPronoun(rawWord, pronounType);
  return paradigm ? getDeclensionPronounFlat(paradigm) : [];
}

function getDeclensionPronounFlat(result: SteenPronounParadigm): string[] {
  if (!result) {
    return [];
  }
  const forms: any = [];
  if (result.cases) {
    forms.push(...Object.values(result.cases));
  }
  if (result.casesSingular) {
    forms.push(...Object.values(result.casesSingular));
  }
  if (result.casesPlural) {
    forms.push(...Object.values(result.casesPlural));
  }
  const dirty: string[] = forms
    .flat()
    .join('/')
    .replace(/ /g, '')
    .split('/')
    .filter(Boolean)
    .reduce((acc: string[], item: string) => {
      if (item.indexOf('(') !== -1 || item.indexOf(')') !== -1) {
        acc.push(item.replace(/[()]/g, ''));
        acc.push(item.replace(/\(.*\)/, ''));
      } else {
        acc.push(item);
      }
      return acc;
    }, []);
  return Array.from(new Set(dirty));
}

export function declensionPronoun(
  rawWord: string,
  pronounType: string,
): SteenPronounParadigm | null {
  // now we don't know how to decline the phrases
  if (rawWord.includes(' ')) {
    return null;
  }

  const word = stripDiacritics(transliterate(rawWord, 'isv-Latn'));
  if (pronounType === 'personal' || pronounType === 'reflexive') {
    if (FIRST_PERSON.includes(word)) {
      return {
        type: 'noun',
        columns: ['singular', 'plural'],
        cases: {
          nom: ['ja', 'my'],
          acc: ['mene (mę)', 'nas'],
          gen: ['mene', 'nas'],
          loc: ['mně', 'nas'],
          dat: ['mně (mi)', 'nam'],
          ins: ['mnojų', 'nami'],
        },
      };
    } else if (SECOND_PERSON.includes(word)) {
      return {
        type: 'noun',
        columns: ['singular', 'plural'],
        cases: {
          nom: ['ty', 'vy'],
          acc: ['tebe (tę)', 'vas'],
          gen: ['tebe', 'vas'],
          loc: ['tobě', 'vas'],
          dat: ['tobě (ti)', 'vam'],
          ins: ['tobojų', 'vami'],
        },
      };
    } else if (THIRD_PERSON.includes(word)) {
      return {
        type: 'adjective',
        casesSingular: {
          nom: ['on', 'ono', 'ona'],
          acc: ['(n)jego (go)', '(n)jego (go)', '(n)jų'],
          gen: ['(n)jego', '(n)jej'],
          loc: ['(n)jem', '(n)jej'],
          dat: ['(n)jemu (mu)', '(n)jej'],
          ins: ['(n)jim', '(n)jejų'],
        },
        casesPlural: {
          nom: ['oni / one', 'one'],
          acc: ['(n)jih / (n)je', '(n)je'],
          gen: ['(n)jih'],
          loc: ['(n)jih'],
          dat: ['(n)jim'],
          ins: ['(n)jimi'],
        },
      };
    } else if (REFLEXIVE.includes(word)) {
      return {
        type: 'noun',
        columns: ['wordForm'],
        cases: {
          nom: [null],
          acc: ['sebę (sę)'],
          gen: ['sebe'],
          loc: ['sobě'],
          dat: ['sobě (si)'],
          ins: ['sobojų'],
        },
      };
    } else if (ALONE.includes(word)) {
      return {
        type: 'adjective',
        casesSingular: {
          nom: ['sam', 'sama', 'samo'],
          acc: ['samogo / sam', 'samų', 'samogo / samo'],
          gen: ['samogo', 'samoj', 'samogo'],
          loc: ['samom', 'samoj', 'samom'],
          dat: ['samomu', 'samoj', 'samomu'],
          ins: ['samym', 'samojų', 'samym'],
        },
        casesPlural: {
          nom: ['sami / same', 'same'],
          acc: ['samyh / same', 'same'],
          gen: ['samyh'],
          loc: ['samyh'],
          dat: ['samym'],
          ins: ['samymi'],
        },
      };
    }
  } else if (pronounType === 'possessive' && POSSESSIVE.includes(word)) {
    return {
      type: 'noun',
      columns: ['wordForm'],
      cases: {
        nom: [rawWord],
        acc: [rawWord],
        gen: [rawWord],
        loc: [rawWord],
        dat: [rawWord],
        ins: [rawWord],
      },
    };
  } else if (
    ['indefinite', 'interrogative', 'negative', 'relative'].includes(
      pronounType,
    ) &&
    (rawWord.includes('čto') ||
      rawWord.includes('kto') ||
      rawWord === 'kogo') &&
    !rawWord.includes('ktory')
  ) {
    let prefix = '';
    let postfix = '';
    let origWord = rawWord;
    if (rawWord === 'kogo') {
      origWord = 'kto';
    } else if (word.match(/koli$/)) {
      postfix = 'koli';
    } else if (word.match(/nebud$/)) {
      postfix = '-nebųď';
    } else if (word.match(/libo$/)) {
      postfix = '-libo';
    } else if (word.match(/^ino/)) {
      prefix = 'ino';
    } else if (word.match(/^ne/)) {
      prefix = 'ně';
    } else if (word.match(/^ni/)) {
      prefix = 'ni';
    } else if (word.match(/^vse/)) {
      prefix = 'vse';
    }
    if (origWord.match(/čto/)) {
      return {
        type: 'noun',
        columns: ['Word Form'],
        cases: {
          nom: [prefix + 'čto' + postfix],
          acc: [prefix + 'čego' + postfix],
          gen: [prefix + 'čego' + postfix],
          loc: [prefix + 'čem' + postfix],
          dat: [prefix + 'čemu' + postfix],
          ins: [prefix + 'čim' + postfix],
        },
      };
    } else if (origWord.match(/kto/)) {
      return {
        type: 'noun',
        columns: ['Word Form'],
        cases: {
          nom: [prefix + 'kto' + postfix],
          acc: [prefix + 'kogo' + postfix],
          gen: [prefix + 'kogo' + postfix],
          loc: [prefix + 'kom' + postfix],
          dat: [prefix + 'komu' + postfix],
          ins: [prefix + 'kym' + postfix],
        },
      };
    } else {
      return null;
    }
  } else if (pronounType === 'relative' && rawWord === 'iže') {
    return {
      type: 'adjective',
      casesSingular: {
        nom: ['iže', 'iže', 'iže'],
        acc: ['(n)jegože', '(n)jegože', '(n)jųže'],
        gen: ['(n)jegože', '(n)jejže'],
        loc: ['(n)jemže', '(n)jejže'],
        dat: ['(n)jemuže', '(n)jejže'],
        ins: ['(n)jimže', '(n)jejųže'],
      },
      casesPlural: {
        nom: ['iže', 'iže'],
        acc: ['(n)jihže / (n)ježe', '(n)ježe'],
        gen: ['(n)jihže'],
        loc: ['(n)jihže'],
        dat: ['(n)jimže'],
        ins: ['(n)jimiže'],
      },
    };
  } else if (
    [
      'demonstrative',
      'indefinite',
      'interrogative',
      'relative',
      'possessive',
      'negative',
      'universal',
    ].includes(pronounType)
  ) {
    const [origWord, postfix] = splitPostfix(rawWord, word);
    const adjectiveParadigm = declensionAdjective(origWord, postfix);
    return {
      type: 'adjective',
      casesSingular: adjectiveParadigm.singular,
      casesPlural: adjectiveParadigm.plural,
    };
  }
  return null;
}

const DEMONSTRATIVE_PRONOUNS: Record<string, string> = {
  ona: 'onȯj',
  ono: 'onȯj',
  onoj: 'onȯj',
  ota: 'otȯj',
  oto: 'otȯj',
  otoj: 'otȯj',
  ova: 'ov',
  ovo: 'ov',
  sa: 'sėj',
  se: 'sėj',
  sej: 'sėj',
  tamta: 'tamtȯj',
  tamto: 'tamtȯj',
  tamtoj: 'tamtȯj',
  tuta: 'tutȯj',
  tuto: 'tutȯj',
  tutoj: 'tutȯj',
  ta: 'tȯj',
  to: 'tȯj',
  toj: 'tȯj',
  te: 'tȯj',
  vsa: 'veś',
  vse: 'veś',
  vsi: 'veś',
};

const DEMONSTRATIVE_POSTFIXES: Record<string, string> = {
  ze: 'že',
  koli: 'koli',
  libo: '-libo',
  nebud: '-nebųď',
};

function splitPostfix(original: string, standard: string) {
  const [match = '', postfix = ''] =
    standard.match(/-?(nebud|koli|libo|ze)$/) || [];
  const main = original.slice(0, -match.length || undefined);

  return [
    DEMONSTRATIVE_PRONOUNS[main] || main,
    DEMONSTRATIVE_POSTFIXES[postfix] || postfix,
  ];
}
