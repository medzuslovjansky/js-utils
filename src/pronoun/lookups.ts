export const FIRST_PERSON = [
  'ja',
  'mene',
  'me',
  'mne',
  'mi',
  'mnoju',
  'my',
  'nas',
  'nam',
  'nami',
];

export const SECOND_PERSON = [
  'ty',
  'tebe',
  'te',
  'tobe',
  'ti',
  'toboju',
  'vy',
  'vas',
  'vam',
  'vami',
];

export const THIRD_PERSON = [
  'on',
  'jego',
  'go',
  'je',
  'jemu',
  'mu',
  'njim',
  'ona',
  'ono',
  'ju',
  'jej',
  'jeju',
  'njeju',
  'oni',
  'one',
  'jih',
  'jim',
  'njimi',
];

export const REFLEXIVE = ['sebe', 'se', 'sobe', 'si', 'soboju'];

export const ALONE = ['sam', 'sama', 'samo', 'sami', 'same'];

export const POSSESSIVE = ['jih', 'jej', 'jego'];
