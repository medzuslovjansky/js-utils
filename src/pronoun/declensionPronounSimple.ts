import { parsePos } from '../partOfSpeech';
import { declensionPronoun } from './declensionPronoun';

export function declensionPronounSimple(lemma: string, morphology: string) {
  const pos = parsePos(morphology);
  if (pos.name !== 'pronoun')
    throw new TypeError(`${lemma} (${morphology} is not a pronoun`);

  return declensionPronoun(lemma, pos.type);
}
