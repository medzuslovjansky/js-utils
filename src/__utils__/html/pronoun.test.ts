/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { declensionPronounSimple } from '../../pronoun/declensionPronounSimple';
import * as fixtures from '../fixtures';
import * as pronoun from './pronoun';

describe('renderers', () => {
  describe('HTML', () => {
    describe('pronoun declension', () => {
      describe('adjective type', () => {
        test('render', () => {
          const to = declensionPronounSimple('to', 'pron.dem.')!;
          expect(pronoun.render(to)).toMatchSnapshot();
        });

        test('renderDiff', () => {
          const before = declensionPronounSimple('to', 'pron.dem.')!;
          const after = declensionPronounSimple('tamto', 'pron.dem.')!;
          expect(pronoun.renderDiff(before, after)).toMatchSnapshot();
        });
      });

      describe('noun type', () => {
        test('render', () => {
          const to = declensionPronounSimple('kto', 'pron.int.')!;
          expect(pronoun.render(to)).toMatchSnapshot();
        });

        test('renderDiff', () => {
          const before = declensionPronounSimple('kto', 'pron.int.')!;
          const after = declensionPronounSimple('čto', 'pron.int.')!;
          expect(pronoun.renderDiff(before, after)).toMatchSnapshot();
        });
      });

      test('should not throw an error given null', () => {
        expect(() => pronoun.render(null)).not.toThrow();
        expect(() => pronoun.renderDiff(null, null)).not.toThrow();
      });

      test('should throw an error when types do not match', () => {
        const before = declensionPronounSimple('to', 'pron.dem.')!;
        const after = declensionPronounSimple('kto', 'pron.int.')!;
        expect(() => pronoun.renderDiff(before, after)).toThrow();
      });

      test('should not throw exceptions', () => {
        const all = [
          ...fixtures.pronouns_demonstrative(),
          ...fixtures.pronouns_indefinite(),
          ...fixtures.pronouns_interrogative(),
          ...fixtures.pronouns_personal(),
          ...fixtures.pronouns_possessive(),
          ...fixtures.pronouns_reciprocal(),
          ...fixtures.pronouns_reflexive(),
          ...fixtures.pronouns_relative(),
        ];

        for (const [word, type] of all) {
          const declension = declensionPronounSimple(word, type);
          expect(() => pronoun.render(declension)).not.toThrow();
          expect(() => pronoun.renderDiff(null, declension)).not.toThrow();
        }
      });
    });
  });
});
