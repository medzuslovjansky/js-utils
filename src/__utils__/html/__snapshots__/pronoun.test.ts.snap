// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderers HTML pronoun declension adjective type render 1`] = `
"<figure>
  <figcaption>Pronoun declension</figcaption>
  <table>
    <caption>singular</caption>
    <thead>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>neuter</th>
        <th>feminine</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>Nom.</th>
        <td>tȯj</td>
        <td>to</td>
        <td>ta</td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td>togo / tȯj</td>
        <td>to</td>
        <td>tų</td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2">togo</td>
        <td>toj</td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2">tom</td>
        <td>toj</td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2">tomu</td>
        <td>toj</td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2">tym</td>
        <td>tojų</td>
      </tr>
    </tbody>
  </table>

  <table>
    <caption>plural</caption>
    <tbody>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>feminine/neuter</th>
      </tr>
      <tr>
        <th>Nom.</th>
        <td>ti / te</td>
        <td>te</td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td>tyh / te</td>
        <td>te</td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2">tyh</td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2">tyh</td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2">tym</td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2">tymi</td>
      </tr>
    </tbody>
  </table>
</figure>
"
`;

exports[`renderers HTML pronoun declension adjective type renderDiff 1`] = `
"<figure>
  <figcaption>Pronoun declension</figcaption>
  <table>
    <caption>singular</caption>
    <thead>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>neuter</th>
        <th>feminine</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>Nom.</th>
        <td><del>tȯj</del> <ins>tamtȯj</ins></td>
        <td><del>to</del> <ins>tamto</ins></td>
        <td><del>ta</del> <ins>tamta</ins></td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td><del>togo / tȯj</del> <ins>tamtogo / tamtȯj</ins></td>
        <td><del>to</del> <ins>tamto</ins></td>
        <td><del>tų</del> <ins>tamtų</ins></td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2"><del>togo</del> <ins>tamtogo</ins></td>
        <td><del>toj</del> <ins>tamtoj</ins></td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2"><del>tom</del> <ins>tamtom</ins></td>
        <td><del>toj</del> <ins>tamtoj</ins></td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2"><del>tomu</del> <ins>tamtomu</ins></td>
        <td><del>toj</del> <ins>tamtoj</ins></td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2"><del>tym</del> <ins>tamtym</ins></td>
        <td><del>tojų</del> <ins>tamtojų</ins></td>
      </tr>
    </tbody>
  </table>

  <table>
    <caption>plural</caption>
    <tbody>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>feminine/neuter</th>
      </tr>
      <tr>
        <th>Nom.</th>
        <td><del>ti / te</del> <ins>tamti / tamte</ins></td>
        <td><del>te</del> <ins>tamte</ins></td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td><del>tyh / te</del> <ins>tamtyh / tamte</ins></td>
        <td><del>te</del> <ins>tamte</ins></td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2"><del>tyh</del> <ins>tamtyh</ins></td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2"><del>tyh</del> <ins>tamtyh</ins></td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2"><del>tym</del> <ins>tamtym</ins></td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2"><del>tymi</del> <ins>tamtymi</ins></td>
      </tr>
    </tbody>
  </table>
</figure>
"
`;

exports[`renderers HTML pronoun declension noun type render 1`] = `
"<figure>
<figcaption>Pronoun declension</figcaption>
<table>
  <thead>
    <tr>
      <th>case</th>
      <th>Word Form</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Nom.</th>
      <td>kto</td>
    </tr>
    <tr>
      <th>Acc.</th>
      <td>kogo</td>
    </tr>
    <tr>
      <th>Gen.</th>
      <td>kogo</td>
    </tr>
    <tr>
      <th>Loc.</th>
      <td>kom</td>
    </tr>
    <tr>
      <th>Dat.</th>
      <td>komu</td>
    </tr>
    <tr>
      <th>Ins.</th>
      <td>kym</td>
    </tr>
  </tbody>
</table>
</figure>
"
`;

exports[`renderers HTML pronoun declension noun type renderDiff 1`] = `
"<figure>
<figcaption>Pronoun declension</figcaption>
<table>
  <thead>
    <tr>
      <th>case</th>
      <th>Word Form</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Nom.</th>
      <td><del>kto</del> <ins>čto</ins></td>
    </tr>
    <tr>
      <th>Acc.</th>
      <td><del>kogo</del> <ins>čego</ins></td>
    </tr>
    <tr>
      <th>Gen.</th>
      <td><del>kogo</del> <ins>čego</ins></td>
    </tr>
    <tr>
      <th>Loc.</th>
      <td><del>kom</del> <ins>čem</ins></td>
    </tr>
    <tr>
      <th>Dat.</th>
      <td><del>komu</del> <ins>čemu</ins></td>
    </tr>
    <tr>
      <th>Ins.</th>
      <td><del>kym</del> <ins>čim</ins></td>
    </tr>
  </tbody>
</table>
</figure>
"
`;
